### PR TITLE
[DNM] Add a Kconfiglib-based menuconfig implementation

### PR DESCRIPTION
**Pull request by ulfalizer**
This do-not-merge pull request adds a Kconfiglib-based `menuconfig` implementation (`menu.py`), built with curses. See the docstring at the top of `menu.py` for an overview. Feedback appreciated.

Run with e.g.

    $ ninja pymenuconfig

Running on Windows requires first installing an appropriate wheel from https://www.lfd.uci.edu/~gohlke/pythonlibs/#curses. There are no dependencies on *nix, as `curses` is in the Python standard library.

Terminal resizing is currently broken on Windows due to an issue with the wheels -- I will work to get that fixed. Everything else should work. Please report issues.

Status
=====

Currently working on some documentation stuff (https://github.com/zephyrproject-rtos/zephyr/pull/7154, https://github.com/zephyrproject-rtos/zephyr/pull/7152, https://github.com/zephyrproject-rtos/zephyr/pull/7152, https://github.com/zephyrproject-rtos/zephyr/issues/7159) while waiting for an update re. the `resize_term()` stuff (https://github.com/zephyrproject-rtos/zephyr/issues/6978#issuecomment-383504596).

That's not a blocker, but maybe we can get that fix in at the same time.

### CODE REVIEW

**Comment by carlescufi** [comment]
Hi @ulfalizer, thanks for the PR! Can you rebase it so we can pull it directly?

**Comment by ulfalizer** [comment]
@carlescufi 
Done

**Comment by carlescufi** [comment]
I get an error when trying this on Windows:

```
C:\Users\Carles\src\zephyr\samples\hello_world\b>ninja pymenuconfig
[0/1] cmd.exe /C "cd /D C:\Users\Carles\src\zephyr\samples\hello_world\b\zephyr\kconfig && "C:\Program Files\CMake\bi...yr/boards/arm/nrf52_pca10040 c:/Users/Carles/src/zephyr//scripts/kconfig/menu.py c:/Users/Carles/src/zephyr//Kconfig"
%1 is not a valid Win32 application
FAILED: CMakeFiles/pymenuconfig
cmd.exe /C "cd /D C:\Users\Carles\src\zephyr\samples\hello_world\b\zephyr\kconfig && "C:\Program Files\CMake\bin\cmake.exe" -E env srctree=c:/Users/Carles/src/zephyr/ KERNELVERSION=1.11.99 KCONFIG_CONFIG=C:/Users/Carles/src/zephyr/samples/hello_world/b/zephyr/.config ENV_VAR_ARCH=arm ENV_VAR_BOARD_DIR=C:/Users/Carles/src/zephyr/boards/arm/nrf52_pca10040 c:/Users/Carles/src/zephyr//scripts/kconfig/menu.py c:/Users/Carles/src/zephyr//Kconfig"
ninja: build stopped: subcommand failed.
```

File: cmake/kconfig.cmake
Comment (carlescufi) [inline cmake/kconfig.cmake]: This should include `${PYTHON_EXECUTABLE}`

**Review by carlescufi — DISMISSED** [review]

**Comment by carlescufi** [comment]
Now that `pymenuconfig` is actually running, I get this:

```
Traceback (most recent call last):
  File "c:/Users/Carles/src/zephyr//scripts/kconfig/menu.py", line 1698, in <module>
    menuconfig(Kconfig("Kconfig" if len(sys.argv) < 2 else sys.argv[1]))
  File "c:/Users/Carles/src/zephyr//scripts/kconfig/menu.py", line 305, in menuconfig
    print(curses.wrapper(_menuconfig))
  File "C:\Python36\lib\curses\__init__.py", line 94, in wrapper
    return func(stdscr, *args, **kwds)
  File "c:/Users/Carles/src/zephyr//scripts/kconfig/menu.py", line 339, in _menuconfig
    c = _menu_win.get_wch()
AttributeError: '_curses.curses window' object has no attribute 'get_wch'
FAILED: CMakeFiles/pymenuconfig
cmd.exe /C "cd /D C:\Users\Carles\src\zephyr\samples\hello_world\b\zephyr\kconfig && "C:\Program Files\CMake\bin\cmake.exe" -E env srctree=c:/Users/Carles/src/zephyr/ KERNELVERSION=1.11.99 KCONFIG_CONFIG=C:/Users/Carles/src/zephyr/samples/hello_world/b/zephyr/.config ENV_VAR_ARCH=arm ENV_VAR_BOARD_DIR=C:/Users/Carles/src/zephyr/boards/arm/nrf52_pca10040 C:/Python36/python.exe c:/Users/Carles/src/zephyr//scripts/kconfig/menu.py c:/Users/Carles/src/zephyr//Kconfig"
ninja: build stopped: subcommand failed.
```

**Comment by ulfalizer** [comment]
@carlescufi 
I'll look into it. In the meantime, try running something like this from the Zephyr root if you just want to try out the interface (stolen from `doc/Makefile`):

```bat
set ENV_VAR_BOARD_DIR=boards\*\*
set ENV_VAR_ARCH=*
set KERNELVERSION=1.9.99
set SRCARCH=x86
python scripts/kconfig/menu.py
```

**Comment by cgohlke** [comment]
What's the "issue with the wheels"?

**Comment by ulfalizer** [comment]
@cgohlke 
Calling `resizeterm()` segfaults -- see https://github.com/zephyrproject-rtos/zephyr/issues/6978#issuecomment-382654223.

Works great other than that. Thanks!

**Comment by cgohlke** [comment]
I did apply https://github.com/wmcbrine/PDCurses/commit/9cf39fe17a1319c85120c88f4bb3cb59fd8739cf last night. Does it not work?

**Comment by ulfalizer** [comment]
@cgohlke 
That's an unrelated issue (thanks for getting the fix in so quickly though). The showstopper is the `resizeterm()` segfault.

I could put together a minimal test case later. I haven't debugged it, so I don't know if calling `resizeterm()` always crashes, or just under specific circumstances.

See https://github.com/zephyrproject-rtos/zephyr/issues/6978#issuecomment-379481024 for some related discussion re. `resizeterm()`. I understand if you don't want to make any larger modifications though, as you're maintaining a million of those wheels.

**Comment by carlescufi** [comment]
Here are my impressions after some time with this:

* Performance is perfectly acceptable on Windows. Not great, but good.
* This implementation in my opinion more than satisfies the requirement for a menuconfig that is cross-platform, allowing us to get rid of the pesky C tools.
* The interface is clean and simple (simpler than `mconf`) and it works reliably
* The resizing on Windows is a problem, as is the fact of having to ask the user to run `python -VV` to find out what Python version and bit-width they are running in order to install the correct wheel version.
* I am willing to live with the above if it's well documented, so the PR should really include the documentation patches
* I believe we should port this to a Python-only toolkit after merging

All in all nice work, a solid foundation.

**Comment by ulfalizer** [comment]
Resizing on Windows is working with the latest curses wheels, but requires a bit of custom hacking for PDCurses. I did some investigation and found a change to the wheel that should allow exactly the same code to be used with ncurses' and PDCurses. It'd just require some extra `resize_term()` calls that would normally be redundant (but harmless) with ncurses: https://github.com/zephyrproject-rtos/zephyr/issues/6978#issuecomment-383504596

I could put together a non-DNM pull request without resizing first, while waiting for feedback on that change.

**Comment by ulfalizer** [comment]
Status
=====

Currently working on some documentation stuff (https://github.com/zephyrproject-rtos/zephyr/pull/7154, https://github.com/zephyrproject-rtos/zephyr/pull/7152, https://github.com/zephyrproject-rtos/zephyr/pull/7152, https://github.com/zephyrproject-rtos/zephyr/issues/7159) while waiting for an update re. the `resize_term()` stuff.

That's not a blocker, but maybe I we can get that fix in at the same time.

**Comment by ulfalizer** [comment]
Superseded by https://github.com/zephyrproject-rtos/zephyr/pull/7174

**Pull request closed without merging**
